spring-context/src/main/java/org/springframework/context/support/DefaultLifecycleProcessor.java

```diff
@@ -143,11 +143,8 @@ private void startBeans(boolean autoStartupOnly) {
 		lifecycleBeans.forEach((beanName, bean) -> {
 			if (!autoStartupOnly || (bean instanceof SmartLifecycle && ((SmartLifecycle) bean).isAutoStartup())) {
 				int phase = getPhase(bean);
-				LifecycleGroup group = phases.get(phase);
-				if (group == null) {
-					group = new LifecycleGroup(phase, this.timeoutPerShutdownPhase, lifecycleBeans, autoStartupOnly);
-					phases.put(phase, group);
-				}
+				LifecycleGroup group = phases.computeIfAbsent(phase,
+						p -> new LifecycleGroup(phase, this.timeoutPerShutdownPhase, lifecycleBeans, autoStartupOnly));
```

Comment:
documentation of `computeIfAbsent` says:

>  return the current (**existing or computed**) value...

so you could do:

```
phases.computeIfAbsent(
						phase,
						p -> new LifecycleGroup(phase, timeoutPerShutdownPhase, lifecycleBeans, autoStartupOnly)
				).add(beanName, bean);
```

But looking at the implementation, `phases` is really just a sorted by Keys, which mean it's really supposed to be a `TreeMap`? So that code could be written as :

```
	private void startBeans(boolean autoStartupOnly) {
		Map<String, Lifecycle> lifecycleBeans = getLifecycleBeans();
		Map<Integer, LifecycleGroup> phases = new TreeMap<>();

		lifecycleBeans.forEach((beanName, bean) -> {
			if (!autoStartupOnly || (bean instanceof SmartLifecycle && ((SmartLifecycle) bean).isAutoStartup())) {

				int phase = getPhase(bean);
				phases.computeIfAbsent(
						phase,
						p -> new LifecycleGroup(phase, timeoutPerShutdownPhase, lifecycleBeans, autoStartupOnly)
				).add(beanName, bean);
			}
		});

		phases.forEach((key, value) -> value.start());
	}
```

The use of `forEach` is safe since it is documented as :

>  Unless otherwise specified by the implementing class, actions are performed in **the order of entry set** iteration

and `TreeMap::entrySet` is documented as :

> The set's iterator returns the entries in ascending key order...

What do you think?

Reply:
I think what you said makes sense, but I added a judgment to reduce the automatic code generation of the iterator.
Thanks !

Reply:
I think what you said makes sense, but I added a judgment to reduce the execution of the iterator automatic code.
Thanks !

Reply:
I'm sorry, but can you explain "but I added a judgment to reduce the execution of the iterator automatic code" - what this means? I don't get it. thank you